pkg/bundle/sync.go

```diff
@@ -441,7 +441,158 @@ func (b *bundle) syncTarget(
 		return false, fmt.Errorf("failed to patch configMap %s/%s: %w", namespace, bundle.Name, err)
 	}
 
-	log.V(2).Info("synced bundle to namespace")
+	log.V(2).Info("synced bundle to namespace for target ConfigMap")
+
+	return true, nil
+}
+
+func (b *bundle) syncSecretTarget(
+	ctx context.Context,
+	log logr.Logger,
+	bundle *trustapi.Bundle,
+	name string,
+	namespace string,
+	data string,
+	shouldExist bool,
+) (bool, error) {
+	secret := &metav1.PartialObjectMetadata{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Secret",
+			APIVersion: "v1",
+		},
+	}
+	err := b.targetCache.Get(ctx, client.ObjectKey{Namespace: namespace, Name: name}, secret)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return false, fmt.Errorf("failed to get Secret %s/%s: %w", namespace, name, err)
+	}
+
+	// If the target obj exists, but the Bundle is being deleted, delete the Secret.
+	if apierrors.IsNotFound(err) && !shouldExist {
+		return false, nil
+	}
+
+	// If the Secret should not exist, but it does, delete it.
+	if !apierrors.IsNotFound(err) && !shouldExist {
+		// apply empty patch to remove the key
+		secretPatch := coreapplyconfig.
+			Secret(name, namespace).
+			WithLabels(map[string]string{
+				trustapi.BundleLabelKey: bundle.Name,
+			}).
+			WithOwnerReferences(
+				v1.OwnerReference().
+					WithAPIVersion(trustapi.SchemeGroupVersion.String()).
+					WithKind(trustapi.BundleKind).
+					WithName(bundle.GetName()).
+					WithUID(bundle.GetUID()).
+					WithBlockOwnerDeletion(true).
+					WithController(true),
+			)
+
+		if err = b.patchSecretResource(ctx, secretPatch); err != nil {
+			return false, fmt.Errorf("failed to patch secret %s/%s: %w", namespace, bundle.Name, err)
+		}
+
+		return true, nil
+	}
+
+	target := bundle.Spec.Target
+	if target.Secret == nil {
+		return false, errors.New("target not defined")
+	}
+
+	// Generated JKS is not deterministic - best we can do here is update if the pem cert has
+	// changed (hence not checking if JKS matches)
+	dataHash := fmt.Sprintf("%x", sha256.Sum256([]byte(data)))
+	targetData := map[string][]byte{
+		target.Secret.Key: []byte(data),
+	}
+
+	if target.AdditionalFormats != nil {
+		if target.AdditionalFormats.JKS != nil {
+			encoded, err := jksEncoder{password: []byte(DefaultJKSPassword)}.encode(data)
+			if err != nil {
+				return false, fmt.Errorf("failed to encode JKS: %w", err)
+			}
+
+			targetData[target.AdditionalFormats.JKS.Key] = encoded
+		}
+
+		if target.AdditionalFormats.PKCS12 != nil {
+			encoded, err := pkcs12Encoder{password: DefaultPKCS12Password}.encode(data)
+			if err != nil {
+				return false, fmt.Errorf("failed to encode PKCS12: %w", err)
+			}
+
+			targetData[target.AdditionalFormats.PKCS12.Key] = encoded
+		}
+	}
```

Comment:
suggestion: I think stuff like this will always be identical for any target, whether ConfigMap or Secret or anything in the future. Ideally this should should be abstracted, and then `syncSecretTarget` and `syncConfigMapTarget` should just apply the generated bundle(s) to targets.

Reply:
Done